### PR TITLE
feat: add opt-in support for defining menuLinks in custom-application-config.json

### PR DESCRIPTION
**Pull request by emmenko**

### CODE REVIEW

File: packages/application-config/schema.json [hunk 1]
```diff
@@ -14,13 +14,6 @@
         "type": "string"
       },
       "uniqueItems": true
-    },
-    "hstsDirective": {
```
Comment (emmenko) [inline packages/application-config/schema.json:18]: Unrelated, but there is no need to keep this as a schema reference. We can simply inline it.

File: packages/application-config/schema.json [hunk 2]
```diff
@@ -158,13 +166,148 @@
...
+          "uniqueItems": true
         }
       },
       "additionalProperties": false,
-      "required": ["csp"]
+      "required": [
+        "csp"
+      ]
+    },
+    "menuLinks": {
```
Comment (emmenko) [inline packages/application-config/schema.json:184]: ⬇️ Here starts the new schema definition for `menuLinks`

File: packages/application-config/src/process-config.ts [hunk 3]
```diff
@@ -12,24 +13,33 @@ import {
   getIsProd,
   getOrThrow,
 } from './utils';
-import substituteEnvVariablePlaceholders from './substitute-env-variable-placeholders';
+import substituteVariablePlaceholders from './substitute-variable-placeholders';
```
Comment (emmenko) [inline packages/application-config/src/process-config.ts:16]: There are now additional variable placeholders, therefore I renamed the file.

File: packages/application-config/src/process-config.ts [hunk 4]
```diff
@@ -102,26 +113,36 @@ const processConfig = ({
...
+                mcApiUrl.host.replace('mc-api', 'mc'),
+              ].join(''),
+              initialProjectKey: appConfig.env.development?.initialProjectKey,
+              teamId: appConfig.env.development?.teamId,
+              oAuthScopes: appConfig.oAuthScopes,
+            })
+          : undefined,
+        menuLinks: appConfig.menuLinks,
+        // @ts-expect-error: the `accountLinks` is not explicitly typed as it's only used by the account app.
+        accountLinks: appConfig.accountLinks,
```
Comment (emmenko) [inline packages/application-config/src/process-config.ts:133]: If defined, we now include the menu links into the `window.app.__DEVELOPMENT__` object.

Note that there is also the `accountLinks`. This is an undocumented/hidden field that is only required for the links of the account app.

File: packages/application-config/src/substitute-variable-placeholders.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+ * NOTE:
+ * Allows variable placeholders. Supported types are:
+ * - `env`: For example `${env:MC_API_URL}`.
+ * - `intl`: For example `${intl:en:Menu.title}`.
+ * - `path`: For example `${path:./app.svg}`, or `${path:@commercetools-frontend/assets/application-icons/heart.svg}`.
+ */
+const variableSyntax = /\${([ ~:\w.'",\-/()@]+?)}/g;
+const envRefSyntax = /^env:/g;
+const intlRefSyntax = /^intl:/g;
+const filePathRefSyntax = /^path:/g;
```
Comment (emmenko) [inline packages/application-config/src/substitute-variable-placeholders.ts:15]: Here you can see the two new placeholder syntax: `intl:` and `path:`

File: packages/application-config/src/substitute-variable-placeholders.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
+import fs from 'fs';
```
Comment (emmenko) [inline packages/application-config/src/substitute-variable-placeholders.ts:1]: The file got renamed, I'll point out only the new things in this file. The rest is the same as before.

File: packages/application-config/src/substitute-variable-placeholders.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+  Boolean(valueOfEnvConfig.match(variableSyntax));
+
+const isEnvVariablePlaceholder = (valueOfPlaceholder: string) =>
+  Boolean(valueOfPlaceholder.match(envRefSyntax));
+
+const isIntlVariablePlaceholder = (valueOfPlaceholder: string) =>
+  Boolean(valueOfPlaceholder.match(intlRefSyntax));
+
+const isFilePathVariablePlaceholder = (valueOfPlaceholder: string) =>
+  Boolean(valueOfPlaceholder.match(filePathRefSyntax));
```
Comment (emmenko) [inline packages/application-config/src/substitute-variable-placeholders.ts:33]: New helpers to determine which function to run.

File: packages/application-config/src/substitute-variable-placeholders.ts [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+      `Missing message key '${requestedIntlMessageId}' specified in config as 'intl:${locale}:${requestedIntlMessageId}'.`
+    );
+  }
+
+  const escapedMatchedString = matchedString.replace(/[${}:]/g, '\\$&');
+  return valueOfEnvConfig.replace(
+    new RegExp(`(${escapedMatchedString})+`, 'g'),
+    translations[requestedIntlMessageId]
+  );
+};
```
Comment (emmenko) [inline packages/application-config/src/substitute-variable-placeholders.ts:83]: This resolves the references of intl messages

File: packages/application-config/src/substitute-variable-placeholders.ts [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+      encoding: 'utf-8',
+    }
+  );
+
+  const escapedMatchedString = matchedString.replace(/[${}:]/g, '\\$&');
+  return valueOfEnvConfig.replace(
+    new RegExp(`(${escapedMatchedString})+`, 'g'),
+    content
+  );
+};
```
Comment (emmenko) [inline packages/application-config/src/substitute-variable-placeholders.ts:108]: This resolves the references of file paths, to load and inline the content as a string.

File: packages/application-config/src/substitute-variable-placeholders.ts [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+              meta
+            );
+          } else if (isFilePathVariablePlaceholder(valueOfPlaceholder)) {
+            substitutedValue = substituteFilePathVariablePlaceholder(
+              valueOfPlaceholder,
+              matchedString,
+              substitutedValue,
+              meta
+            );
+          }
```
Comment (emmenko) [inline packages/application-config/src/substitute-variable-placeholders.ts:147]: Each if-else checks for one of the supported placeholder syntax.

File: packages/application-config/src/validate-config.ts [hunk 11]
```diff
@@ -10,7 +10,7 @@ type ErrorAdditionalProperty = ErrorObject<
 >;
 type ErrorEnum = ErrorObject<'enum', { allowedValues: string[] }>;
 
-const ajv = new Ajv({ strict: true });
+const ajv = new Ajv({ strict: true, useDefaults: true });
```
Comment (emmenko) [inline packages/application-config/src/validate-config.ts:13]: This allows the schema validator to inject default values to the config, according to the schema.

File: packages/application-config/test/fixtures/config-file-path-variables.json [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+  "name": "avengers-app",
+  "entryPointUriPath": "avengers",
+  "cloudIdentifier": "gcp-eu",
+  "env": {
+    "production": {
+      "url": "https://my-app.com"
+    }
+  },
+  "menuLinks": {
+    "icon": "${path:@commercetools-frontend/assets/images/camera.svg}",
```
Comment (emmenko) [inline packages/application-config/test/fixtures/config-file-path-variables.json:11]: Here you can see how the `path` placeholder works.

File: packages/application-config/test/fixtures/config-intl-variables.json [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+      "url": "https://my-app.com"
+    }
+  },
+  "menuLinks": {
+    "icon": "<svg><path fill=\"#000000\" /></svg>",
+    "defaultLabel": "Avengers",
+    "labelAllLocales": [
+      {
+        "locale": "en",
+        "value": "${intl:en:Menu.Avengers}"
```
Comment (emmenko) [inline packages/application-config/test/fixtures/config-intl-variables.json:16]: Here you can see how the `intl` placeholder works

File: packages/application-config/test/fixtures/config-oidc.json [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+  },
+  "oAuthScopes": {
+    "view": [
+      "view_orders",
+      "view_states"
+    ],
+    "manage": [
+      "manage_orders"
+    ]
+  }
```
Comment (emmenko) [inline packages/application-config/test/fixtures/config-oidc.json:21]: Unrelated but the fixture/test was missing before.

File: packages/application-shell/src/components/application-shell/redirect-to-login.tsx [hunk 15]
```diff
@@ -31,7 +31,7 @@ const generateAndCacheNonceWithState = (state: AuthorizeSessionState) => {
 const RedirectToLogin = () => {
   const location = useLocation();
 
-  if (window.app.__DEVELOPMENT__) {
+  if (window.app.__DEVELOPMENT__?.oidc?.authorizeUrl) {
```
Comment (emmenko) [inline packages/application-shell/src/components/application-shell/redirect-to-login.tsx:34]: The shape of the `__DEVELOPMENT__` field changed a bit (see constants file). Now the OIDC config is scoped under `oidc` field.

File: packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.ts [hunk 16]
```diff
@@ -1,60 +1,252 @@
...
+              permissions: menuLink.permissions,
+              // @ts-ignore: not defined in schema, as it's only used internally.
+              featureToggle: menuLink.featureToggle ?? null,
+            }))
+          : [],
+    },
+  };
+};
+
 function useApplicationsMenu<Key extends MenuKey>(
```
Comment (emmenko) [inline packages/application-shell/src/hooks/use-applications-menu/use-applications-menu.ts:227]: The implementation of the `useApplicationsMenu` changed a bit now:

* it supports loading the menu links from the `window.app.__DEVELOPMENT__.menuLinks`.
* it uses the Apollo cache to "inject" the menu links and let the Apollo query load the data only from the cache (development only).

File: packages/assets/application-icons/heart.svg [hunk 17]
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 0 24 24" width="24px" fill="#000000"><path d="M16 3.7c-1.7 0-3.3.9-4.1 2.4-.9-1.5-2.4-2.4-4-2.4-2.5 0-4.6 2.1-4.8 5 0 .1-.1.8.2 1.9.4 1.6 1.2 3 2.5 4.1l6.2 5.7 6.3-5.7c1.2-1.1 2.1-2.6 2.5-4.1.3-1.1.2-1.8.2-1.9-.4-3-2.5-5-5-5z"/></svg>
```
Comment (emmenko) [inline packages/assets/application-icons/heart.svg:1]: All the icons here are/will be used for org-level custom apps.

File: playground/src/components/entry-point/entry-point.js [hunk 18]
```diff
@@ -69,9 +69,6 @@ const EntryPoint = () => (
         globalActions.handleActionError(error, 'sdk')(dispatch);
     }}
     applicationMessages={loadMessages}
-    DEV_ONLY__loadNavbarMenuConfig={() =>
-      import('../../../menu.json').then((data) => data.default || data)
-    }
```
Comment (emmenko) [inline playground/src/components/entry-point/entry-point.js:74]: 🙌

**Review by islam3zzat — APPROVED** [review]
Thank you 🙌

File: lint-staged.config.js [hunk 19]
```diff
@@ -7,7 +7,20 @@ module.exports = {
...
+    //
+    //   No tests found related to files changed since last commit.
+    //   Run Jest without `-o` or with `--all` to run all tests.
+    //   Error: An error occurred while adding the reporter at path "/path/to/file".Reporter is not a constructor
+    //
+    // For that reason, we move the `--onlyChanged` flag next to it.
+    'yarn lint:js --reporters=jest-silent-reporter --onlyChanged',
+    'tsc-files --noEmit',
+  ],
+  '*.{ts,tsx}': [
```
Comment (pa3) [inline lint-staged.config.js:23]: Isn't `'*.{js}'` now exactly the same as the ` '*.{ts,tsx}'` config? Why splitting?

File: packages/application-config/test/process-config.spec.js [hunk 20]
```diff
@@ -486,6 +491,201 @@ describe('processing a config with environment variable placeholders', () => {
   });
 });
 
+describe('processing a config with intl variable placeholders', () => {
+  beforeEach(() => {
+    loadConfig.mockReturnValue(fixtureConfigIntlVariables);
+  });
+  it('should process the config and correctly parse the translations', () => {
+    const result = processConfig(createTestOptions());
+    expect(result).toEqual({
```
Comment (pa3) [inline packages/application-config/test/process-config.spec.js:500]: I know snapshots have a bad rep nowadays, but `toMatchInlineSnapshot` does a quite decent job in cases like this one. =)

**Review by pa3 — APPROVED** [review]

File: lint-staged.config.js (hunk 19 above)
Comment (emmenko) [reply lint-staged.config.js:23]: Using `tsc-files` on JS files does not really work. So by splitting it we can run `tsc-files` only on actual TS files.

But I actually fixed that in another PR: https://github.com/commercetools/merchant-center-application-kit/pull/2313/files#diff-29c15f0b5d8d1144bd22153c302cde16c2de092d65b122a618cbdb3023336346

Let me revert the change here to not mix things up. Thanks for bringing this up.

File: packages/application-config/test/process-config.spec.js (hunk 20 above)
Comment (emmenko) [reply packages/application-config/test/process-config.spec.js:500]: Yeah inline snapshots are actually useful in such cases.

**Pull request merged**